Fixes #16327 - "Use Latest" is incorrectly pointing to lower version

kgaikwad opened the pull request:
During selection of puppet module - "Use Latest" is incorrectly pointing to lower version than available in the repo.
Use Latest should point to actual Latest version available.
With this commit, actual Latest version will appear on a list as well so that Use Latest will point to correct version.

bbuckingham commented:
@kgaikwad, Thanks for the PR!

Observation while testing out the changes, with a puppet modules with 2 versions (2.0.1 and 1.3.0)
- scenario 1:
  - add the 2.0.1 version to the content view
  - click 'Select new version'
  - observe 3 rows ("Use latest (currently 2.0.1)', 2.0.1 and 1.3.0)

- scenario 2:
  - add the 1.3.0 version to the content view
  - click 'Select new version'
  - observe 2 rows ("Use latest (currently 2.0.1)' and 2.0.1)

My assumption is that in both cases, there should only be 2 rows displayed to the user.  1 for Latest and one for the version that was not selected.

bbuckingham commented on app/controllers/katello/api/v2/content_views_controller.rb:
```diff
@@ -93,11 +93,18 @@ def show
+        query = query.where(:name => params[:name])
+        if current_uuids.present?
+          module_by_name = current_cv_puppet_modules.find_by(:name => params[:name])
+          current_uuids.delete(module_by_name.uuid) if module_by_name && module_by_name.latest_in_modules_by_author?(query)
```
Can you briefly describe the scenario (or example) that requires the additional query logic and deletion of record from the result set?

kgaikwad replied:
Take example, there is one puppet module called "abc" having below versions:

name  version author
----------------------
abc   4.0.0   xyz
abc   3.0.0   xyz
abc   2.0.0   xyz
abc   1.0.2   pqr
abc   1.0.1   pqr
abc   1.0.0   pqr

With existing code,  User selected "4.0.0" version for content-view "test-cv". When he again clicked "select version", he will able to see below list:

Use Latest (currently 3.0.0)
abc   3.0.0   xyz
abc   2.0.0   xyz
Use Latest (currently 1.0.2)
abc   1.0.2   pqr
abc   1.0.1   pqr
abc   1.0.0   pqr

This "available_puppet_modules" API serves the results with & without name parameter for content-view.
"Use Latest" computation logic is totally present on client side in file [content-view.factory.js](engines/bastion_katello/app/assets/javascripts/bastion_katello/content-views/content-view.factory.js).

In this case, on the client side its  difficult to get latest version for selected puppet module as each version contains its own version value in computed version field and at this point don't have additional info about currently selected puppet module version.

Instead of skipping selected puppet module version which is latest module version among all versions of puppet module, I have added it in result list i.e by deleting the version from current_uuids list.

For verification whether this selected puppet module version is latest or not, I have added these additional queries.
By doing this, "Use latest" points to the latest version.

Does this make sense or should I change the result of this API on server side so that I get required information to client side for "Use Latest" computation?

bbuckingham replied:
@kgaikwad, Thank you for the detailed explanation.  The new API logic appears to be what is causing the behavior that I mentioned in my other comment.

The following what I observed:
- If I select version 2.0.1,
  - with the change, available_puppet_modules is returning 2.0.1 and 1.3.0 (Unexpected extra row in the UI)
  - without the change, available_puppet_modules is returning 1.3.0 (UI behavior Ok)

- If I select version 1.3.0,
  - with the change, available_puppet_modules is returning 2.0.1 (UI behavior Ok)
  - without the change, available_puppet_modules is returning 2.0.1 (UI behavior Ok)

- If I select version 'use latest 2.0.1',
  - with the change, available_puppet_modules is returning 2.0.1 and 1.3.0 (UI behavior Ok)
  - without the change, available_puppet_modules is returning 2.0.1 and 1.3.0 (UI behavior Ok)

The ideal solution, would probably be for the 'available_puppet_modules' to take in to account the 'latest' and not return the puppet module details when it is selected.  E.g. In my case, if i select 'Use Latest (currently 2.0.1)', it would not return 2.0.1.  That said, the UI may likely also require modifications to that it can still indicate to the user the appropriate information.

kgaikwad replied:
@bbuckingham ,

Sorry for repeating the same thing, but could you please confirm, my understanding on below scenarios is correct or not according to your points.

My modifications in code should be in a such a way that,

1) If I select version 'use latest 2.0.1 and click 'select version' then it will show 3 rows ->

a. Use Latest (currently 2.0.1)
b. 2.0.1 
c. 1.3.0
 
2) If I select version '2.0.1' and click 'select version' then it will show 2 rows ->

a. Use Latest (currently 2.0.1)
b. 1.3.0

3) If I select version '1.3.0' and click 'select version' then it will show 2 rows ->

a. Use Latest (currently 2.0.1)
b. 2.0.1

bbuckingham replied:
@kgaikwad, I believe what you mention is more what a user would expect.  The difference that I observed was that scenario 2) would return 1.3.0 and 2.0.1 in the API response.  This would then result in 3 rows in the UI (similar to scenario 1)).

kgaikwad commented:
@bbuckingham , sorry for delay in updating the pull request.

I have made correction on client side so that "Use Latest" will point to latest version of selected puppet module list and it will not add extra row for latest version.

For above example of puppet module with 2 versions (2.0.1 and 1.3.0) 

1.  when added the 2.0.1 version to the content view 
     It will show only 2 rows ("Use latest (currently 2.0.1)' and 1.3.0) on 'Select new version' click

2. when added "Use Latest (currently 2.0.1)"  to the content view 
    It will show only 2 rows ("Use latest (currently 2.0.1)' and 1.3.0) on 'Select new version' click

3. when added the 1.3.0 version to the content view 
   It will show only 1 row  "Use latest (currently 2.0.1)'  on 'Select new version' click

Please let me know if any changes required.

bbuckingham commented:
@kgaikwad, Thanks for the updates!  I apologize for the delay.  I missed your posting just before the shutdown. 

Unfortunately, we need to tweak the behavior.  Below are the basic scenarios; however, I have done a ~~strikethrough~~ where the current behavior is different than I expected.

1. When user goes in to initially add the puppet module, the following should be listed for available versions:
    a. Use Latest (currently 2.0.1)
    ~~b. 2.0.1~~  This entry is missing.
    c. 1.3.0

2. If user selects version 'Use Latest (currently 2.0.1)' and click 'select new version' then it will show 3 rows ->
    a. Use Latest (currently 2.0.1)
    ~~b. 2.0.1~~ This entry is missing.
    c. 1.3.0

~~3. If user selects version '2.0.1' and click 'select new version' then it will show 2 rows ->
    a. Use Latest (currently 2.0.1)
    b. 1.3.0~~  The user does not currently have the option to select 2.0.1.

4. If user selects version '1.3.0' and click 'select new version' then it will show 2 rows ->
    a. Use Latest (currently ~~2.0.1~~)  It should show 2.0.1, but currently being listed as 1.3.0.
    ~~b. 2.0.1~~  This entry is missing.

Note: The thing to be aware of with respect to 'Latest', when a user selects it, the UUID stored in the DB (ContentViewPuppetModule) for the module is 'nil'.  The reason is that 'Latest' can change as the user uploads new versions of the module.   That way, if I upload 3.0, my CV will get it the next time it is published without me having to go explicitly select 3.0.  This is the reason we show both 'Use Latest (Currently 2.0.1)' and '2.0.1').

kgaikwad commented:
@bbuckingham ,
Thank you for being patient and explaining me the scenario in brief. 
I apologize for making so much confusion. Now, I made the changes as you mentioned and removed client-side 'Use Latest' code logic and moved it into controller as on client-side I don't have list of selected modules.
One thing I am not sure about how locale for 'Always Use Latest' will get handled?
or Should I mention respective key and value in all locale files?

Please let me know if any other changes required.

bbuckingham commented:
@kgaikwad, Thanks for the updates.  I apologize for the delay. We have an issue with the UI that was making it a bit of a bear to test and I wanted to ensure that it was not related to this PR.  For reference, that issue is: http://projects.theforeman.org/issues/17922

The scenarios do look better now.  :)

Regarding the locale files, you do not need to update them directly.  As long as the logic is defined properly (e.g. using _("")) and it appears to be, the translations will be taken care of separately.

bbuckingham commented on app/controllers/katello/api/v2/content_views_controller.rb:
```diff
@@ -92,17 +92,28 @@ def show
     param :id, :identifier, :desc => N_("content view numeric identifier"), :required => true
     param :name, String, :desc => N_("module name to restrict modules for"), :required => false
     def available_puppet_modules
```
Given the extra logic added to available_puppet_modules and add_use_latest_records, would it be possible to add a couple of controller tests?

kgaikwad commented:
@bbuckingham ,
Thank for reviewing the pull-request. I also noticed the same issue you mentioned after rebase with current changes. As Walden is doing changes on Web-UI side, I thought it is effect of that so not pointed it while updating the pull-request. I will check on it.

As per you mentioned, I will add test cases for extra logic added by me in controller and will update the pull-request as earliest.

kgaikwad commented:
@bbuckingham ,

I have added test cases for new changes added by me on controller side. I have tried to resolve @houndci-bot's wanings but it is giving me error while running the tests.

Tried fixing by making below changes -
process :available_puppet_modules, :method => :get, :params => { :id => @library_dev_staging_view.id, :name => "trystero" }

which is giving below errorwhile running respective test - 
'ActionController::UrlGenerationError: No route matches {:action=>"available_puppet_modules", :controller=>"katello/api/v2/content_views"}'

Please give me some pointers.

bbuckingham commented:
[test]

bbuckingham commented:
@kgaikwad, I was able to run rubocop cleanly locally with the PR, which appears to be what houndci was complaining about.  In discussion with others in irc, it sounds like there have been some instances where houndci is not reporting properly.  There is also an email on theforeman-dev related to that topic.

So given the clean local run, we will ignore the error above.  Also, the code you added for the test is very consistent with other tests in how it calls the 'get', so +1 from me. :)

bbuckingham commented:
@kgaikwad, ACK from me, once we have a clean test run through jenkins.

I added the [test] to request it to be queued up in jenkins; however, there may be a bit of a delay due to an infrastructure issue.

Also, in order to have your PRs automatically tested, you will need to be in the Katello organization.  @jlsherrill has sent you an invitation.  Once accepted, you will also need to make your org membership public.  See https://help.github.com/articles/publicizing-or-hiding-organization-membership for details.

bbuckingham commented:
[test]

bbuckingham commented:
@kgaikwad, the jenkins tests ran clean.  Thank you for your contribution and persistence on this PR!  Merging.. :)
